ui/components/src/Table/VirtualizedTable.tsx

```diff
@@ -0,0 +1,179 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { Column, HeaderGroup, Row, flexRender } from '@tanstack/react-table';
+import { Box } from '@mui/material';
+import { TableVirtuoso, TableComponents, TableVirtuosoHandle, TableVirtuosoProps } from 'react-virtuoso';
+import { useRef, useMemo } from 'react';
+import { TableRow } from './TableRow';
+import { TableBody } from './TableBody';
+import { InnerTable } from './InnerTable';
+import { TableHead } from './TableHead';
+import { TableCell, TableCellProps } from './TableCell';
+import { VirtualizedTableContainer } from './VirtualizedTableContainer';
+import { TableProps } from './model/table-model';
+import { useVirtualizedTableKeyboardNav } from './hooks/useVirtualizedTableKeyboardNav';
+
+type TableCellPosition = {
+  row: number;
+  column: number;
+};
+
+export type VirtualizedTableProps<TableData extends Record<string, unknown>> = Required<
+  Pick<TableProps<TableData>, 'height' | 'width' | 'density'>
+> & {
+  onRowClick: (id: string) => void;
+  rows: Array<Row<TableData>>;
+  columns: Array<Column<TableData, unknown>>;
+  headers: Array<HeaderGroup<TableData>>;
+};
+
+// Separating out the virtualized table because we may want a paginated table
+// in the future that does not need virtualization, and we'd likely lay them
+// out differently.
+export function VirtualizedTable<TableData extends Record<string, unknown>>({
+  width,
+  height,
+  density,
+  onRowClick,
+  rows,
+  columns,
+  headers,
+}: VirtualizedTableProps<TableData>) {
+  const virtuosoRef = useRef<TableVirtuosoHandle>(null);
+
+  // Use a ref for these values because they are only needed for keyboard
+  // focus interactions and setting them on state will lead to a significant
+  // amount of unnecessary re-renders.
+  const visibleRange = useRef({
+    startIndex: 0,
+    endIndex: 0,
+  });
+  const setVisibleRange: TableVirtuosoProps<TableData, unknown>['rangeChanged'] = (newVisibleRange) => {
+    visibleRange.current = newVisibleRange;
+  };
+
+  const keyboardNav = useVirtualizedTableKeyboardNav({
+    visibleRange: visibleRange,
+    virtualTable: virtuosoRef,
+
+    // We add 1 here for the header.
+    maxRows: rows.length + 1,
+    maxColumns: columns.length,
+  });
+
+  const getFocusState = (cellPosition: TableCellPosition): TableCellProps['focusState'] => {
+    if (cellPosition.row === keyboardNav.activeCell.row && cellPosition.column === keyboardNav.activeCell.column) {
+      return keyboardNav.isActive ? 'trigger-focus' : 'focus-next';
+    }
+
+    return 'none';
+  };
+
+  const VirtuosoTableComponents: TableComponents<TableData> = useMemo(() => {
```

Comment:
I've done my best to hit some obvious perf things with `useMemo` and `useCallback` in some obvious places. I'll continue to take passes on this as I'm working on the table legend, but I think I got all the big ones to start.